src/Http/WebUtilities/src/MultipartSectionPipeReader.cs

```diff
@@ -0,0 +1,360 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.WebUtilities
+{
+    internal class MultipartSectionPipeReader : PipeReader
+    {
+        private readonly PipeReader _pipeReader;
+        private bool _finished = false;
+        private bool _metadataSkipped = false;
+        private readonly MultipartBoundary _boundary;
+        private int _partialMatchIndex = 0;
+        private ReadOnlySequence<byte> _buffer;
+        private SequencePosition _endPosition;
+        private bool _isReaderComplete;
+
+        private static ReadOnlySpan<byte> CrlfDelimiter => new byte[] { (byte)'\r', (byte)'\n' };
+        private static ReadOnlySpan<byte> EndOfFileDelimiter => new byte[] { (byte)'-', (byte)'-' };
+
+        public bool CanSeek { get; }
+
+        public long RawLength { get; private set; } = 0;
+
+        public long? LengthLimit { get; private set; }
+        public long Length { get; private set; } = 0;
+
+        internal MultipartSectionPipeReader(PipeReader pipeReader, MultipartBoundary boundary, bool canSeek)
+        {
+            _pipeReader = pipeReader;
+            _boundary = boundary;
+            CanSeek = canSeek;
+        }
+
+        private void UpdateLength(long read)
+        {
+            Length += read;
+            RawLength += read;
+            if (LengthLimit.HasValue && Length > LengthLimit.GetValueOrDefault())
+            {
+                throw new InvalidDataException($"Multipart body length limit {LengthLimit.GetValueOrDefault()} exceeded.");
+            }
+        }
+
+        public override Stream AsStream(bool leaveOpen = false)
+        {
+            return new PipeReaderStream(this, Length, leaveOpen);
```

Comment:
Why did you override this? To set Length? That's not very useful when Length is a changing value. It's probably 0 in most scenarios where AsStream is called (e.g. before reading anything from the body).

Reply:
I think you are right, there isn't much use to Length if you can't seek anyway.